python_spike/caseprocessor/decorators/transaction_decorators.py

```diff
@@ -0,0 +1,84 @@
+from functools import wraps
+from typing import Callable
+import logging
+from db.db_utility import Session
+from sqlalchemy.exc import SQLAlchemyError
+
+'''
+Example callback function signature:
+
+@pubsub_transaction
+@sqlalchemy_transaction
+def callback(message: pubsub_v1.subscriber.message.Message, session: Session) -> None:
+    ... 
+
+'''
+ 
+def pubsub_transaction(sql_transaction: Callable) -> Callable:
+    
+    @wraps(sql_transaction)
+    def with_pubsub_error_handling(*args, **kwargs):
+        pubsub_message = args[0]
+
+        try:
+            sql_transaction(*args, **kwargs)
+            pubsub_message.ack()
+            logging.info(f"Message acked: {pubsub_message.data}")
+
+        except Exception as e:
+            logging.error(f"PubSub decorator caugth error: {e}")
+            pubsub_message.nack()
+            logging.info(f"Message nacked: {pubsub_message.data}\n")
+
+    return with_pubsub_error_handling
+
+def sqlalchemy_transaction(callback: Callable) -> Callable:
+
+    @wraps(callback)
+    def with_transaction_handling(*args, **kwargs):
+        local_session = Session()
+        try:
+            callback(*args, **kwargs, session=local_session)
+            local_session.commit()
+            Session.remove()
+            logging.info("Session commited succesfully!")
+
+        except Exception as e:
+            Session.remove()
+            logging.error(f"DB transaction decorator caught exception: {e}\n")
+            raise Exception("SQL transaction error")
+
+    return with_transaction_handling
+
+
+'''
+Example callback function signature:
+
+@transaction
+def callback(message: pubsub_v1.subscriber.message.Message, session: Session) -> None:
+    ... 
+
+'''
+
+def transaction(callback: Callable) -> Callable:
+
+    @wraps(callback)
+    def with_transaction_handling(*args, **kwargs):
+        local_session = Session()
+        pubsub_message = args[0]
```

Comment:
Where it has to interact with the wrapped functions arguments like this, it's probably better to stick to kwargs rather than positional args

Reply:
That's what I was thinking, but I wasn't able to find what the keyword argument would be! 

If you look at the [docs](https://cloud.google.com/python/docs/reference/pubsub/latest/google.cloud.pubsub_v1.subscriber.client.Client#google_cloud_pubsub_v1_subscriber_client_Client_subscribe) for the subscribe function and scroll down to the parameters list, you'll see that the callback is passed as an argument to the subscribe function and that the callback function "...receives the message as its only argument...", but there's no more details and the argument name isn't stated.
This suggests the PubSub message is somehow passed under the hood to the callback when it arrives and in my understanding will always be the first argument.

I think that the risk of confusion here is relatively low because we'll only pass the callback to the subscribe function, without explicitly passing any arguments to it.
Also my bad, should've probably left a word on that. Let me know what you think.

Reply:
Makes sense, I forgot our design here is also constrained by what the Pubsub library requires for a callback.